### DISCUSSION TITLE
Not possible to use different configurations in staging + production??

### DISCUSSION THREAD

**Discussion by JVMartin** [vercel/next.js#22243] [Help]
Evidently, it's not possible to use different configurations in staging + production.

We have a NextJS application packaged up in Docker, so that it's environment agnostic. Therefore, we need to supply different configurations for `staging` and `production` environments, of course. For instance, our `staging` Next front-end should hit a different back-end `API_URL` than our `production` Next front-end.

[Per the documentation](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration), we should be able to do runtime configuration like this in our `next.config.js`:
```javascript
serverRuntimeConfig: {
  apiUrl: process.env.API_URL,
},
```

If this were true, we would be all set!

But the documentation is definitely wrong, because [per the documentation elsewhere](https://nextjs.org/docs/basic-features/environment-variables):
![image](https://user-images.githubusercontent.com/4450814/108138673-52584a80-7073-11eb-83fb-477ea0013ae6.png)

So since `process.env.*` is always interpolated at **build time** into the compiled assets, it appears to not be possible to use different configurations in different environments, totally defeating the purpose of ***environment*** variables.

This totally violates all principles of "build once, deploy anywhere", "immutable deployments", "test the artifact before promoting to production", etc.

Does Next deliberately violate these principles or is this an accident? Any ideas of any hacks or workarounds to let me run a Next application in multiple environments - *without* multiple build artifacts, because it is obviously critical that the artifact being promoted into production be identical to the one in test / staging environments?

Relevant issues, all totally unresolved:
https://github.com/vercel/next.js/issues/1488
https://github.com/vercel/next.js/discussions/15131
https://stackoverflow.com/questions/60364413/react-next-js-recommended-way-to-set-constants-such-as-backend-api-urls
https://stackoverflow.com/questions/60871862/nextjs-prevent-env-vars-to-be-required-on-build-time
https://github.com/vercel/next.js/discussions/17641

**Comment by leerob** [comment]
Is there any reason not to use Environment Variables for this?

**Reply by JVMartin** [reply]
Environment variables would not work, because [per the documentation](https://nextjs.org/docs/basic-features/environment-variables):
![image](https://user-images.githubusercontent.com/4450814/108138673-52584a80-7073-11eb-83fb-477ea0013ae6.png)

So since `process.env.*` is always interpolated at **build time** into the compiled assets, it appears to not be possible to use different configurations in different environments, totally defeating the purpose of ***environment*** variables.

**Reply by leerob** [reply]
Maybe if we step back - your original problem is you want to use different configurations for staging + production. What's the use case for reading values at _runtime_ for different configurations instead of at build time to differentiate environments?

E.g. if you're in preview, use the preview env vars to connect to one database. If you're in prod, use the other set.

**Reply by JVMartin** [reply]
As I said in my post, you should never use different compiled code in different environments. That would totally violate all proper software engineering principles: "build once, deploy anywhere", "immutable deployments", "test the artifact before promoting to production", etc. If the artifacts you're deploying to production aren't *literally the same* artifacts you've promoted through staging and test environments, you are doing it wrong.

**Reply by JVMartin** [reply]
@leerob Anything interpolated at static time *cannot* be changed at runtime, so there's literally no way to use different vars in different environments when you've done the interpolation statically.

**Reply by dlindenkreuz** [reply]
@leerob I‘d love to read a follow-up on this topic – could you give this another look? 👀

**Reply by pkit** [reply]
It doesn't work with env variables too. Mainly because all env vars are "cached" in `.next/required-server-files.json` (wtf is that?)
So subsequent `next build`s with **different** env vars still produce the same result.
Need to delete `.next` dir between builds. Who thought that it's a good idea?

**Reply by CleberRossi** [reply]
@JVMartin and @pkit  I'm facing the same issue. 
However, reading the documentation I think it is possible to use, some worthy mention points.
You cannot use PUBLIC env because NEXT will replace it. (Need to use next.config.js)
You cannot use static generation.

Other thing is that I tested my .next folder to check if process.env were being changed, it seems NOT. It only happened for PUBLIC variables. So you can do it, but you cannot use static generation again. 

I will be creating a docker file and try to use this approach to check. Some other posts to help you out:
https://github.com/vercel/next.js/issues/22266
https://blog.vulcanjs.org/how-to-set-configuration-variables-in-next-js-a81505e43dad
https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

Conclusion for now: If you are using server side rendering it is possible to use process.env

**Reply by pkit** [reply]
I don't see any clear steps or explanation here.
I don't care about SSR.
I do care that env vars I use, will be used at build time. That's it.

**Reply by dtinth** [reply]
> Maybe if we step back - your original problem is you want to use different configurations for staging + production. What's the use case for reading values at _runtime_ for different configurations instead of at build time to differentiate environments?

@leerob Consider a case where a web-based application is packaged and distributed as a Docker image (e.g. for on-premises installation). Popular open source examples include [n8n](https://docs.n8n.io/hosting/installation/docker/), [pgAdmin](https://www.pgadmin.org/docs/pgadmin4/latest/container_deployment.html), [Redis Commander](https://hub.docker.com/r/rediscommander/redis-commander), and [Mongo Express](https://hub.docker.com/_/mongo-express). These applications all allow users to configure the behavior of the app via environment variables, _without having to rebuild an image._

For this kind of app, we want to let users easily run it by just doing `docker run --env-file=.env <image>`

It is unreasonable to ask the user to download the source code and build the Docker image by themselves each time they want to change the configuration. First, it is extremely inconvenient. And second, in SaaS companies that offer on-premises installation, we want to ship the compiled package to clients and want to keep the source code private.

Recent features in Next.js has worsened the situation. [Output File Tracing and App Directory has broken the Runtime Configuration feature](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) with no documented alternative. Right now I’m using these [workarounds](https://notes.dt.in.th/NextRuntimeEnv), but it’d be great if Next.js offers a first-class solution for this, or at least, document in the [Docker deployment docs](https://nextjs.org/docs/deployment#docker-image) on how to allow a single build artifact to be reused in multiple environments.

**Comment by michrome** [comment]
Hey @JVMartin `serverRuntimeConfig` does read from the environment as per the [runtime configuration documentation](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration). Try the following:

1. Add the following to `next.config.js`.

```js
serverRuntimeConfig: {
  mySecret: process.env.MY_SECRET,
}
```

2. Create a new page.

```js
import getConfig from "next/config";
const { serverRuntimeConfig } = getConfig();

function Page({ secret }) {
  return <div>My secret: {secret}</div>;
}

Page.getInitialProps = async (ctx) => {
  const secret = serverRuntimeConfig.mySecret;
  return { secret: secret };
};

export default Page;
```

3. Create a `.env` file.

```env
MY_SECRET=foobar
```

4. Run `npx next build`. _Note: you will see `{"mySecret":"foobar"}` included in `required-server-files.json`_.

5. Run `npx next start` and view your page in a browser.

<img width="578" alt="Screenshot 2021-02-18 at 21 06 16" src="https://user-images.githubusercontent.com/596770/108421958-d6eec880-722d-11eb-831b-4ab0f21ed0b3.png">

6. Stop the server, update `.env` with a new value, re-run `npx next start` (don't rebuild!) and refresh the browser.

```env
MY_SECRET=new value
```

<img width="578" alt="Screenshot 2021-02-18 at 21 08 39" src="https://user-images.githubusercontent.com/596770/108421987-e0783080-722d-11eb-9bf9-bef3dac0d522.png">

7. Stop the server, delete the `.env` file, run `MY_SECRET="another value" npx next start` (again, don't rebuild) and refresh the browser.

<img width="578" alt="Screenshot 2021-02-18 at 21 10 27" src="https://user-images.githubusercontent.com/596770/108422014-e837d500-722d-11eb-823b-33557f84cf2a.png">

**Reply by brycefranzen** [reply]
For what it is worth. I work at the same organization as @JVMartin and we ultimately ended up using the [next.js runtime configuration documentation](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) to solve our issue.  

The way the site builds is still odd though-- it builds with whatever env params are provided at build-time (they're hard-coded in the build output); however, when using the "runtime config" helpers, it magically works and disregards the env params in the build output and uses the params passed in.

Basically, do what @michrome said, and it'll work just fine.

**Reply by michrome** [reply]
> Basically, do what @michrome said, and it'll work just fine.

If only my kids would sign up to this thinking…

**Comment by geekcubed** [comment]
Has there been any new information on this?  We're faced with the exact same issue.

Our NextJS product is built and baked into a Docker image.  This is then ready to be distributed to our customers to install on their own infrastructure, often inside their network firewall.  Our app requires a bunch of stuff to be configured during / after deployment: Graph API connection strings, Azure Active Directory configuration etc.  Which means we're facing the same challenges: configuration stored in either `.env` files or `.next.config.js` is baked into the compiled code during build, even if we're using `process.env.*`.  

Specifically for us, a lot of this configuration code is consumed "outside" of a page - e.g. in a hook or library object, I'm not even sure we can we even apply the `getInitalProps` hack (and that's what it is)?  Right now it seems to be less work to dump NextJS and go back to vanilla React backed by a more production-ready backend

**Reply by AgentChris** [reply]
Hey, @geekcubed @JVMartin we faced a similar issue (for configuring sentry environment property)
What we did was to create a config endpoint and this is how we solved it. I know is not ideal but at least we respect the "build once, deploy anywhere"

_.env_
```
NEXT_PUBLIC_ENVIRONMENT=DEVELOP
```
_next.config_
```
...
  publicRuntimeConfig: {
    environment: process.env.NEXT_PUBLIC_ENVIRONMENT,
  },
  ...
```
_pages/config.ts_
```
import { NextApiRequest, NextApiResponse } from 'next';
import getConfig from 'next/config';

export default async function handler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
  res.status(200).json(getConfig());
}
```
_sentry.client.config_
```
import { NextConfig } from 'next';
import * as Sentry from '@sentry/nextjs';
import axios, { AxiosResponse } from 'axios';

const SENTRY_DSN = process.env.SENTRY_DSN || process.env.NEXT_PUBLIC_SENTRY_DSN;

axios.get('/api/config').then((response: AxiosResponse<NextConfig>) => {
  const { publicRuntimeConfig } = response.data;
  Sentry.init({
    dsn: SENTRY_DSN,
    // Adjust this value in production, or use tracesSampler for greater control
    tracesSampleRate: 1.0,
    environment: publicRuntimeConfig?.environment,
  });
});
```
_sentry.server.config_
```
import * as Sentry from '@sentry/nextjs';
import getConfig from 'next/config';

const SENTRY_DSN = process.env.SENTRY_DSN || process.env.NEXT_PUBLIC_SENTRY_DSN;

const { publicRuntimeConfig } = getConfig();
Sentry.init({
  dsn: SENTRY_DSN,
  // Adjust this value in production, or use tracesSampler for greater control
  tracesSampleRate: 1.0,
  environment: publicRuntimeConfig.environment,
});
```

**Reply by geekcubed** [reply]
[We]'ve ended up just using the `process.env` props and really restricted where we consume the settings and trusting that it works.  For now, we've gotten away with it, but it's far from ideal and I don't know enough about what's happening under the hood to appreciate what we've compromised by doing it this way.

With background in php and .net, I'm struggling to see the benefits that next brings to the party in terms of server-side rendering

**Reply by stigkj** [reply]
I'm not sure about the reason for having the env.vars. getting built in like Next.js is doing, but there is at least an easy way around it: use another library for reading config, like node-config.

**Reply by dlindenkreuz** [reply]
@AgentChris You should probably not serve the config object as-is. It might contain secrets meant only for the server.
```typescript
export default async function handler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
  const { publicRuntimeConfig } = getConfig(); // ← 👀 
  res.status(200).json(publicRuntimeConfig);
}
```

**Comment by mnebuerquo** [comment]
The root problem is that when you build the content statically, you have to choose what content to build. If you have staging content and production content, you can't "build once, deploy anywhere" with static page generation at build time. You either build with staging content or you build with production content. This is true regardless of how you package your build artifacts. You can use docker or tar files or anything, but static content in the image makes it environment specific.

The choice seems to me as either of the following: 
1. Build separate stage and production images with everything built statically in the image, or 
2. Build an environment agnostic image which can generate static content at startup.

**The "build once, deploy anywhere" philosophy is at odds with having statically generated pages and content built into the image. There isn't a way to do both, regardless of how you import your configuration.** 

This isn't a question of "doing it wrong" or not following best practices. The decisions made by the Next team are optimized for the specific feature of generating static content at build-time and keeping secret configuration from landing in the browser in the process. If you want "build once, deploy anywhere" as the priority, then you would make different decisions in the platform architecture. There wouldn't be a build-time option, instead opting to build on-demand or at server startup.

**Reply by JVMartin** [reply]
Everyone at my organization is running Next as an express server, doing server-side rendering, taking data and transforming it to pages server-side in real-time. We are not using it as a static site builder, we would use a more appropriate tool like Gatsby for that. So, Next definitely should, in our use case, conform to every aspect of a 12-factor app, and should definitely follow the **build once, deploy anywhere** philosophy.

**Reply by hongymagic** [reply]
Unironically, this is the way the commercials are for Vercel. More builds = more workers = $50USD/month (per extra worker). Also serves the "serverless" I-charge-you-per-API-hit ecosystem very well.

I guess in this paradigm, no one cares that code mutates between environment to environment....

**Reply by AnthonyDugarte** [reply]
As of now, you don't need to build all of your site on build time, and you can make usage of static generation across environments, with a single build, by making usage of [ISR](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration), I think it is a nice feature that allows you to fix this particular issue.

**Reply by tjcampbell** [reply]
@AnthonyDugarte that unfortunately isn't true. Anything statically generated at build time will include injected runtime config that was specified at build time. For dynamic routes you can defer static generation, but not for other pages that depend on runtime config. So the initial page load for every environment will be the same, until cache invalidation where pages are regenerated with the the new runtime config.

**Comment by volodymyr-strilets-mindcurv** [comment]
Check this lib for the client-side environments 
https://github.com/andrewmclagan/react-env#readme

**Comment by dtinth** [comment]
The [Runtime Configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) feature is now [considered “legacy”](https://github.com/vercel/next.js/pull/43443) and do not work with [Output File Tracing](https://nextjs.org/docs/advanced-features/output-file-tracing) (they gets inlined during build) or [React Server Components](https://nextjs.org/docs/advanced-features/react-18/server-components) (crashes the dev server).

I’ve written a note about this topic: [**Runtime Environment Variables in Next.js**](https://notes.dt.in.th/NextRuntimeEnv)

What I currently do right now:

- [In server-side code (API routes and `getServerSideProps`) I use `global.process.env` instead of `process.env`.](https://notes.dt.in.th/NextRuntimeEnv#api-routes-use-global-process-env)
- [In client-side-only components I just do client-side fetching using React Query (via tRPC) to fetch public variables from an API route that exposes the public variables.](https://notes.dt.in.th/NextRuntimeEnv#client-side-rendered-components-fetch-the-data-you-need-from-an-api)

**Reply by fireai-developer** [reply]
Thanks @dtinth for spelling it out so clearly. I ended up at the same approach for client-side-only components as you, so glad to see I was not the only one thinking in that direction. 👍

**Reply by lucasbasquerotto** [reply]
I think that a client-side request to fetch the variables is a no go if you need SSR/SSG, and the variables need to be set beforehand. It could be possible fetching in the server side, before anything else, although I hadn't tried yet.

**Comment by dangduytung** [comment]
To solve for the problem about multiple environments in building Docker of the application NextJs, I wrote the answer here:
[https://stackoverflow.com/questions/77329575/different-env-variables-for-the-same-nextjs-docker-image/78112804#78112804](https://stackoverflow.com/questions/77329575/different-env-variables-for-the-same-nextjs-docker-image/78112804#78112804)

**Comment by aliaz0** [comment]
How can I modify the contents of the `next.config.js` file itself to handle two different environments? I need to set different hosts for `images.remotePatterns` based on the environment. Is it possible to build a single Docker image and use it across multiple environments, each with a different Next.js configuration?
To clarify, I’m not referring to runtime variables used in the application (which I already manage with the [`next-runtime-env`](https://www.npmjs.com/package/next-runtime-env) package) but specifically to changing the Next.js configuration at runtime, in any way possible.

**Reply by lucasbasquerotto** [reply]
I'm not using `remotePatterns`, but I'm using environment variables in other parts of the config file. It should be possible using `process.env.VAR_NAME`:

```js
images: {
    remotePatterns: process.env.REMOTE_PATTERNS
}
```

Then, you just need to specify `REMOTE_PATTERNS` as the docker container environment variable.

**Reply by aliaz0** [reply]
@lucasbasquerotto Are you sure about this? If I use `process.env.VAR` in my Next.js configuration, the build process fails with the error: `"images.remotePatterns[0].hostname" is missing`. If I provide these environment variables on the machine where the Docker image is built, they get resolved to their values at build time and cannot be changed at runtime. Am I misunderstanding something?

**Reply by lucasbasquerotto** [reply]
@aliaz0 Make sure to follow the structure required for `remotePatterns`. For example, in this case it might need to be:

```js
images: {
    remotePatterns: [ hostname: process.env.REMOTE_PATTERN_HOSTNAME ]
}
```

and possibly other fields that may be required.

About the second part of your comment, NextJS should use the current environment variables that are defined in the config file when running the container, like it happens with the variables in the field `publicRuntimeConfig`, but it might be that places that you define them may use them only during build and they end up "hardcoded" in the generated build artifacts. 

Building in a container with environment variable `MY_VAR: "A"` and running in a container with `MY_VAR: "B"` should populate the config file with `MY_VAR: "B"` when running NextJS, but if the variable is defined in a setting that is used only during build, only the build setting will be used, unfortunately. I don't know if this is the case of `remotePatterns` (running a test of this case should be able to provide the answer).

**Comment by tappin-kr** [comment]
Would be nice to get an answer/guidance for this question.

**Reply by JVMartin** [reply]
There still is no clear way to do it. NextJS doesn't want to give you any clear guidance on deploying via docker images because that doesn't give money to Vercel.

btw it's super easy to have different vars in different envs with Vercel. Funny that

**Comment by sefin23** [comment]
This is a well-known pain point and has been discussed extensively. Here's a consolidated summary of the working approaches as of 2024/2025:

## The Core Problem

`NEXT_PUBLIC_*` variables are **inlined at build time** into JS bundles — you cannot change them post-build. Server-only `process.env.*` variables (without the `NEXT_PUBLIC_` prefix) are **read at runtime** on the server, so they work fine for SSR/API routes. The issue is client-side code.

## Working Solutions

### 1. For server-side code only (SSR, API Routes, Server Components)
Just use `process.env.MY_VAR` — it's read at runtime on the server. No tricks needed. This alone may solve most use cases if you're doing SSR.

In App Router Server Components:
```js
// This is fine — read at runtime, never bundled into client JS
const apiUrl = process.env.API_URL;
```

### 2. Use `next-runtime-env` package (recommended for client-side)
The [`next-runtime-env`](https://www.npmjs.com/package/next-runtime-env) package (mentioned by @aliaz0) is the cleanest solution for making runtime env vars available on the client without baking them into the build:

```bash
npm install next-runtime-env
```

It works by generating a `__env.js` script served at runtime that exposes variables to the client. Your Docker image stays environment-agnostic.

### 3. Expose public vars via an API route (no extra package)
As shown by @AgentChris and @dtinth — create an API route that reads `publicRuntimeConfig` or `process.env` server-side and returns it to the client:

```ts
// pages/api/config.ts
import getConfig from 'next/config';
export default function handler(req, res) {
  const { publicRuntimeConfig } = getConfig();
  res.status(200).json(publicRuntimeConfig);
}
```

Then fetch this on the client before initializing things that need the config (e.g., Sentry, analytics). **Important:** Only expose `publicRuntimeConfig`, never `serverRuntimeConfig`, to avoid leaking secrets.

### 4. Use `global.process.env` for server-side code (workaround for Output File Tracing)
As noted by @dtinth, `serverRuntimeConfig` is now **legacy** and broken with Output File Tracing and Server Components. In server-side code, you can bypass the build-time substitution using:

```js
const apiUrl = global.process.env.API_URL; // not substituted at build time
```

### 5. For `next.config.js` options like `remotePatterns`
These are **build-time only** — `process.env` in `next.config.js` is read when the config is evaluated during build. You cannot change these at runtime without rebuilding. For Docker use cases, consider building with placeholder values and post-processing, or keep image domain rules permissive and validate elsewhere.

## Summary Table

| Use case | Solution |
|---|---|
| Server-side vars (SSR/API) | `process.env.VAR` — works at runtime |
| Client-side runtime vars | `next-runtime-env` package |
| Client-side without package | API route exposing `publicRuntimeConfig` |
| App Router server components | `process.env.VAR` or `global.process.env.VAR` |
| `next.config.js` settings | Build-time only — no runtime solution |

The **"build once, deploy anywhere"** pattern is fully achievable for SSR-based Next.js apps — the limitation only applies to statically generated pages that bake config into HTML/JS at build time.

**Reply by aliaz0** [reply]
The question is about `next.config.js`, and still there is no solution for that.
